File: 0046-permutations/0046-permutations.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */

const permute = nums => {
    const dfs = (startIndex, used, path, res) => {
        if (startIndex === nums.length) {
            res.push([...path]);
            return;
        }
        
        for (let i = 0; i < nums.length; i++) {
            if (used[i]) continue;
            path.push(nums[i]);
            used[i] = true;
            dfs(startIndex + 1, used, path, res);
            path.pop();
            used[i] = false;
        }
    }
    const res = [];
    dfs(0, new Array(nums.length).fill(false), [], res);
    return res;
};

File: 0046-permutations/0046-permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        def dfs(index, path, res, nums):
            if index == len(nums):
                res.append(path.copy())
                return
            for num in nums:
                if num not in path:
                    path.append(num)
                    dfs(index + 1, path, res, nums)
                    path.pop()
        
        res = []
        dfs(0, [], res, nums)
        return res